[블랙잭 1단계] 모찌 미션 제출합니다.

wondroid-world opened the pull request:
# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [X] 프로그램이 정상적으로 작동하는가?
- [X] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [X] 코딩 스타일 가이드를 준수하였는가?
    - [X] IDE 코드 자동 정렬을 적용하였는가?
    - [X] 린트 검사를 통과하였는가?
- [X] 프로그래밍 요구 사항을 준수하였는가?
- [X] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

1. **객체가 메시징을 던진다.** 
이 의미를 어떻게 판단해야 되는 지 고민을 해봤습니다. 처음에는 객체는 메시징을 던진다라는 말을 들었을 때, 다른 객체에서 완전히 원하는 포맷의 주어야한다고 생각을 했습니다. 

```
 fun getCardsInfomation(): String = value.map { it.combine().joinToString() }
```

카드들의 정보를 출력할 때 List<String> 형태가 아닌 OutputView에서 출력하기 편한 String 타입으로 반환하는 것이 더 좋다고 생각했습니다. 하지만, 객사오를 읽으면서 객체는 자율적으로 행동한다라는 문장을 읽으면서 완전히 상대방이 원하는 값을 주는 것이 아니라, 상태에 대한 메시징을 던지는 것이 중요하다고 생각했습니다.

```
 fun getCardsInfomation(): List<String> = value.map { it.combine() }
```

그래서 위와 같이 수정을 했습니다. 메시징을 던진다에 대해서 저의 생각의 변화에 대해서, 페로로의 생각이 궁금해요!

2. **어디까지 확장성을 생각해야할까요**

저는 설계를 할 때 확장성을 많이 고려하는 편입니다. 카드를 생성할 때도 모든 카지노에서 쓸 수 있도록 아래와 같이 구현했습니다.

```
package blackjack.model

enum class Denomination(
    val title: String,
    val number: Int,
) {
    ACE("A", 1),
    TWO("2", 2),
    THREE("3", 3),
    FOUR("4", 4),
    FIVE("5", 5),
    SIX("6", 6),
    SEVEN("7", 7),
    EIGHT("8", 8),
    NINE("9", 9),
    TEN("10", 10),
    JACK("J", 11),
    QUEEN("Q", 12),
    KING("K", 13),
}

```

하지만, 이렇게 확장성을 고려하다보니, 블랙젝에서 구현하는 데, 블랙잭 점수에 대해서 너무 많은  분기를 처리해야하더라구요. 그래서, 아래와 같이 블랙잭에 핏한 카드들을 생성했습니다.

```
package blackjack.model

enum class Denomination(
    val title: String,
    val number: Int,
) {
    ACE("A", 0),
    TWO("2", 2),
    THREE("3", 3),
    FOUR("4", 4),
    FIVE("5", 5),
    SIX("6", 6),
    SEVEN("7", 7),
    EIGHT("8", 8),
    NINE("9", 9),
    TEN("10", 10),
    JACK("J", 10),
    QUEEN("Q", 10),
    KING("K", 10),
}

```

일단 현실에 주어진 것을 구현한 다음, 확장성을 고려해야할까요? 둘 중 어떤게 선행이 되어야 하는 지 궁금합니다.

wondroid-world committed:
docs: 기능 목록 작성

m6z1 committed:
docs: 기능 요구 사항 추가

- 카드 모양 추가

m6z1 committed:
feat: 카드 모양 객체 구현

m6z1 committed:
feat: 카드 객체 구현

- 카드는 모양과 끗수를 가진다.

m6z1 committed:
docs: 기능 요구 사항 추가

- 카드의 끗수에 대한 요구 사항 추가

m6z1 committed:
feat: 카드 끗수를 숫자로 변환하는 로직 구현

m6z1 committed:
feat: 플레이어 행동 객체 구현

- 응답에 대한 플레이어 행동 반환

wondroid-world committed:
Merge branch 'step1' of https://github.com/wondroid-world/kotlin-blackjack into step1

wondroid-world committed:
docs: 플레이어 상태 기능 목록 수정

wondroid-world committed:
feat: 플레이어 객체 구현

wondroid-world committed:
feat: 플레이어 카드 추가 기능 구현

wondroid-world committed:
docs: 플레이어 기능 구현 추가

wondroid-world committed:
feat: 플레이어 이름만 가질 경우, 비어있는 카드 리스트를 가짐

m6z1 committed:
feat: 게임 결과 상태 객체 구현

m6z1 committed:
docs: 게임 결과 기능 목록 수정

m6z1 committed:
feat: 기준 점수와 비교 점수를 비교하여 기준 점수 기준으로 게임 결과 반환 로직 구현

m6z1 committed:
feat: 딜러 객체 구현

- 이름, 카드 리스트 가짐

m6z1 committed:
docs: 딜러 이름 기능 목록 추가

m6z1 committed:
feat: 딜러 이름의 초기값은 '딜러'

m6z1 committed:
docs: 딜러 카드 리스트 기능 목록 추가

m6z1 committed:
feat: 딜러 카드 리스트 초기값 구현

m6z1 committed:
feat: 딜러 카드 추가 기능 구현

wondroid-world committed:
docs: 카드들 상태 기능 목록 수정

wondroid-world committed:
feat: 게임 진행 중 카드들의 상태 구현

wondroid-world committed:
docs: 끗수 기능 목록 추가

wondroid-world committed:
feat: 끗수 객체 추가

wondroid-world committed:
docs: 카드 덱 기능 목록 추가

wondroid-world committed:
feat: 카드 생성기 객체 구현

wondroid-world committed:
docs: 카드 덱 기능 목록 수정

wondroid-world committed:
feat: 카드 덱 객체 구현

wondroid-world committed:
docs: 카드들 기능 목록 수정

wondroid-world committed:
feat: 카드들 객체 구현

wondroid-world committed:
refactor: 카드 생성자 끗수를 끗수 타입으로 변경

wondroid-world committed:
feat: 끗수 숫자 추가

wondroid-world committed:
refactor: ktlint 수정

wondroid-world committed:
refactor: 카드의 함수 네이밍 변경

wondroid-world committed:
docs: 플레이어 이름 검증 로직 추가

wondroid-world committed:
refactor: ktlint 수정

wondroid-world committed:
feat: 딜러 이름을 가지도록 수정

wondroid-world committed:
docs: 카드들 합 기능 목록 수정

m6z1 committed:
feat: 카드 점수 계산 로직 구현

m6z1 committed:
refactor: 플레이어와 딜러의 카드들 타입 변경 및 카드 추가 로직 수정

- 기존) List<Card>
- 변경) Cards

m6z1 committed:
docs: 플레이어 행동 기능 목록 수정

m6z1 committed:
feat: 딜러 점수에 따른 플레이어 행동 반환 로직 구현

m6z1 committed:
feat: 카드 상태 반환 로직 구현

m6z1 committed:
feat: 플레이어 카드 입출력 로직 구현

m6z1 committed:
feat: dsl 실습 구현

m6z1 committed:
docs: DSL 노트 정리 수정

m6z1 committed:
feat: 입출력 구현

m6z1 committed:
feat: 카드들의 블랙잭, 버스트 반환 로직 추가

m6z1 committed:
feat: 딜러 결과 반환 및 업데이트 로직 추가

m6z1 committed:
feat: 게임 결과의 반댓값 반환 로직 구현

m6z1 committed:
feat: 플레이어의 블랙잭 여부 및 결과 업데이트 로직 추가

m6z1 committed:
feat: 플레이어 이름 검증 로직 추가

m6z1 committed:
feat: 블랙잭 로직 구현

- 딜러의 hit 과 stay 로직 제외 구현

wondroid-world committed:
feat: 딜러의 점수 분기 로직 추가

wondroid-world committed:
feat: 플레이어 이름 검증 로직 추가

m6z1 committed:
refactor: 컨틀롤러 내부의 함수 순서 변경

m6z1 committed:
refactor: 사용하지 않는 생성자 제거

m6z1 committed:
refactor: 카드들의 파라미터 네이밍 변경

m6z1 committed:
test: 카드들 테스트 추가

m6z1 committed:
refactor: 플레이어에서 bust, blackjack 판단하도록 수정

m6z1 committed:
test: 카드 테스트 추가

m6z1 committed:
test: 딜러 테스트 추가

m6z1 committed:
test: 게임 결과 테스트 추가

m6z1 committed:
test: 플레이어 테스트 추가

wondroid-world committed:
homework: DSL 추가

laco-dev commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,10 @@
+) {
+    fun isDenominationAce(card: Card): Boolean = card.denomination == Denomination.ACE
+
+    fun combine(): String = denomination.title + shape.koreanName
```
> 객체가 메시징을 던진다.
이 의미를 어떻게 판단해야 되는 지 고민을 해봤습니다. 처음에는 객체는 메시징을 던진다라는 말을 들었을 때, 다른 객체에서 완전히 원하는 포맷의 주어야한다고 생각을 했습니다.

기본적으로 객체가 꽁꽁 숨겨져있고, 객체로 하여금 필요한 행동을 메시징 이라는 형태로 요청한다고 봐주시면 됩니다.
그렇기에 객체가 스스로 할 수 있도록 자율성을 부여할 수 있게 됩니다.

> 그래서 위와 같이 수정을 했습니다. 메시징을 던진다에 대해서 저의 생각의 변화에 대해서, 페로로의 생각이 궁금해요!

다만 위 예시는 메시징을 통해 가져오는 것과 어울리지 않습니다.
왜냐하면 어떻게 출력할 지 결정하는 것은. Card로부터 필요한 기능이 아니기 때문인데요.

가령 `2하트` 라고출력하는 것을 화면에 따라 `❤️2` 이렇게 출력할지도 모릅니다.
혹은 그림으로 출력할수도 있죠.

메시징을 던지기 이전의 관심사 분리가 선행되어야 할 것으로 보입니다!

wondroid-world replied:
1. **CardShape에서 koreanName 삭제**

[ 관련 커밋 ] 71ed45f6ad97cae304f90b9f71206bc3ba1c5ba0

[ 이유 ]

아! 그럼 페로로의 의미는 카드의 정보만 주면 되고 그것을 어떻게 사용할지는 view의 관심사라고 말하는 걸까요?

페로로, 페로로의 말을 들으면서, 생각을 한것이 있습니다. 지금 카드 모양을 아래와 같이 정의하고 있습니다. 

```
package blackjack.model

enum class CardShape(
    val koreanName: String,
) {
    DIAMOND("다이아몬드"),
    SPADE("스페이드"),
    HEART("하트"),
    CLOVER("클로버"),
}

```

이 블랙잭 게임에서 중요한 게 카드 모양이 뭐가 있는 지만 중요하지, 굳이 koreanName이라는 view에서 쓰는 내용을 정의하지 않아야겠다는 생각이 듭니다. 그리고 페로로 말대로 저 코리안 네임이 모양으로 변경될 수도 있구요! 

재사용성을 위해서는 관심사 분리가 정말 중요한 부분이네요!

wondroid-world replied:
2. **combine메서드 삭제** 

[ 관련 커밋 ] ff80fba180cdc78fdd0f21be13bf33163fba96a4

[ 이유 ]
combine이라는 것이 view를 위해서 만들어진 메서드라서, 메시징이라고 보기어렵다는 판단이 듭니다. 제가 생각하는 메시징은 객체와 객체가 대화하기 위해서 만든 메서드라고 생각합니다. 그래서, 저의 판단에서는 combine은 메시징이라고 판단이 되지 않습니다.

wondroid-world replied:
3. **Denomination의 title 삭제**

[ 관련 커밋 ] 99e50572ecbae1776280f6d58586588ab926e73e

[ 이유 ]
Denomination의 title을 삭제한 이유는 view를 위한 로직이였습니다. domain이 view에 의존하고 있다는 생각이 들었습니다. 그리고 현재 title의 A가 Ace로 변할 수 있다고 생각하여, 블랙잭에서 변할 수 없는 게임 규칙인 number만 남기고 삭제했습니다.

laco-dev commented on src/main/kotlin/blackjack/model/Card.kt:
```diff
@@ -0,0 +1,10 @@
+    val shape: CardShape,
+    val denomination: Denomination,
+) {
+    fun isDenominationAce(card: Card): Boolean = card.denomination == Denomination.ACE
```
자기 스스로가 Card인데 파라미터로 Card를 받는건 무엇인가요?

wondroid-world replied:
1. **isDenominationAce 파라미터 삭제와 함수명 변경**

[ 관련 커밋 ] 9bfc22839a061efe8bdaef70bcebdabe0b229b6f

[ 이유 ] 
자신(카드)의 끗수가 에이스인지 확인하는 로직인데, card를 받고 있습니다. 이를 해결하기 위해서, 파라미터 삭제와 더 직관적인 isAce로 메서드 변경했습니다.

laco-dev commented on src/main/kotlin/blackjack/model/Denomination.kt:
```diff
@@ -0,0 +1,20 @@
+    val title: String,
+    val number: Int,
+) {
+    ACE("A", 0),
+    TWO("2", 2),
+    THREE("3", 3),
+    FOUR("4", 4),
+    FIVE("5", 5),
+    SIX("6", 6),
+    SEVEN("7", 7),
+    EIGHT("8", 8),
+    NINE("9", 9),
+    TEN("10", 10),
+    JACK("J", 10),
+    QUEEN("Q", 10),
+    KING("K", 10),
```
> 하지만, 이렇게 확장성을 고려하다보니, 블랙젝에서 구현하는 데, 블랙잭 점수에 대해서 너무 많은 분기를 처리해야하더라구요. 그래서, 아래와 같이 블랙잭에 핏한 카드들을 생성했습니다.

이미 enum에 점수를 넣은 것 부터 점수를 결정하는 규칙을 도메인에서 결정하였다고 보입니다.
점수가 동적으로 결정될 것을 고민하는 확장성이라면 애초에 enum에 포함되는 것 자체가 모순이겠죠?

다만 지금 블랙잭 규칙에서는 그러한 요구사항이 없기 때문에 enum에 포함되어도 괜찮다고 봅니다.
확장성을 고려하는 것은 중요하지만 과한 확장성은 객체의 복잡도를 크게 높이거나 오버 엔지니어링이 될 위험이 있습니다.

wondroid-world replied:
페로로덕분에 제가 어디까지 생각하고 구현해야하는 지 가닥이 잡히네요! 과한 확장성보다 지금에 주어진 것에 충실해야겠다는 생각이 들어요!

고마워요! 페로로*

laco-dev commented on src/main/kotlin/blackjack/model/CardShape.kt:
```diff
@@ -0,0 +1,10 @@
+enum class CardShape(
+    val koreanName: String,
+) {
+    DIAMOND("다이아몬드"),
+    SPADE("스페이드"),
+    HEART("하트"),
+    CLOVER("클로버"),
```
마찬가지로 카드 모형이 위와 같이 출력될 것이라는 것을 결정하는 것은 도메인이 아닌 UI 의 관심사입니다. 😓

laco-dev commented on src/main/kotlin/blackjack/model/PlayerBehavior.kt:
```diff
@@ -0,0 +1,23 @@
+enum class PlayerBehavior(
+    private val answer: String,
+) {
+    HIT("Y"),
+    STAY("N"),
+    ;
+
+    companion object {
+        fun from(answer: String): PlayerBehavior =
+            when (answer) {
+                "Y" -> HIT
+                "N" -> STAY
+                else -> throw IllegalArgumentException("ddd")
+            }
+
+        fun from(dealerScore: Int): PlayerBehavior {
+            if (dealerScore <= 16) return HIT
+            return STAY
+        }
+    }
```
이어서 플레이어가 어떻게 행동할지 또한, "Y" "N"으로 결정되는 것이 아닌 Hit Stay로 결정되어야 합니다.
만약 화면에서 진행하는 게임이었다면 유저는 "Y"라고 타이핑 하지 않고 버튼을 클릭합니다.

도메인 모델이 UI 패키지에 의존만 하지 않으면 괜찮다는 뜻이 아니에요.
UI와 관심사를 나누는게 근본적인 목적입니다.

wondroid-world replied:
1. **PlayerBehavior의 answer 필드 삭제와 HIT, STAY 구분 로직 inputview로 변경**

[ 관련 커밋 ] a67c225a19a0d9b67acc0f84fbcf5c16360ffb8b

[ 이유 ]
도메인이 view에 의존적이여서, answer이라는 필드 삭제, 구분 로직은 input으로 옮겨 inputView에서 답을 얻을 수 있도록 변경

laco-dev commented on src/main/kotlin/blackjack/model/PlayerBehavior.kt:
```diff
@@ -0,0 +1,23 @@
+                else -> throw IllegalArgumentException("ddd")
+            }
+
+        fun from(dealerScore: Int): PlayerBehavior {
+            if (dealerScore <= 16) return HIT
+            return STAY
+        }
```
한편으로는 딜러 점수가 16점보다 낮으면 HIT라는 결과는 플레이어 행동이 결정할 책임일까요?
아니면 딜러가 스스로 판단할 책임일까요?

wondroid-world replied:
1. **딜러 스스로가 hit or stay할지 판별**

[ 관련 커밋 ] 
dbb7ad8cee13fb929071e223568111e3a7560aed

[ 이유 ] 
페로로, 저는 enum class를 사용하는 이유가 값을 나열하기 위해서 사용한다고 생각해요. 여기서는 블랙잭의 규칙을 나열하기 위해서, enum class를 사용했어요. 그래서, 경기 규칙이 딜러의 상태를 변화시켜야 한다고 생각했습니다. 

그런데, 저의 현재 domain들을 생각해봤을 때, 딜러는 player를 상속받고 있습니다. 그 말은 딜러도 player처럼 hit or stay를 할것인지 자유롭게 판단하는 객체라는 사실입니다. 

이 구조를 바탕으로 생각 했을 때, 딜러가 hit or stay를 하는 것은 딜러의 자율성(책임)에 맡겨야한다고 생각합니다.

이렇게 도메인 구조에 따라, hit or stay하는 책임이 누군지 달라진다고 생각합니다.

이에 대해서 페로로의 생각은 어떨까요?

laco-dev commented on src/main/kotlin/blackjack/model/Cards.kt:
```diff
@@ -0,0 +1,33 @@
+class Cards(
+    value: List<Card>,
+) {
+    private val _value: MutableList<Card> = value.toMutableList()
+    val value: List<Card> get() = _value.map { card -> card.copy() }
```
`card.copy()` 라는 코드는 왜 필요한가요?

wondroid-world replied:
일차적으로 mutable 주소값을 노출하지 않기위해서, 딥카피를 해서, 값을 전달하고 싶었습니다!

laco-dev commented on src/main/kotlin/blackjack/model/CardsStatus.kt:
```diff
@@ -0,0 +1,21 @@
+    companion object {
+        const val BLACKJACK_SCORE = 21
+
+        fun from(
+            cardsScore: Int,
+            firstTurn: Boolean = false,
+        ): CardsStatus {
+            if (firstTurn && cardsScore == 21) return BLACKJACK
+            if (cardsScore > 21) return BUST
+            return NONE
+        }
```
카드의 상태를 결정하는 것은 카드가 스스로 해야 하지 않을까요?
만약 카드 상태를 결정하는데 필요한게 늘어날수록 이 함수도 비대해지게 됩니다.

아래 시그니처와 비교하면 어떤가요?
```kotlin
class Cards {
   ...
   fun getStatus(): CardsStatus { ... }
}
```

wondroid-world replied:
1. **카드의 상태는 카드 스스로가 판단한다.**

[ 관련 커밋 ]
a7eb2678a816366deb43d038f393e2bfe83e5aa0

[ 이유 ]

처음 코드는 카드 상태가 자신의 상태가 무엇인지 판단을 하도록 코드를 작성했습니다. 

지금 생각해보면, 카드의 상태는 카드가 가지는 속성 중 일부라고 판단을 하였습니다. 카드 스스로가 자신의 상태를 판별할 수 있는 로직을 구현했습니다. 이러니, player에서 중복적으로 가진 isBalckjack, isBust 코드도 삭제하여, 다 카드 내에서 판단할 수 있도록 했습니다. 

"객체 스스로 자율적으로 일을 한다." 라는 관점에서 player들이 카드가 isblackjack인지 isBust인지 판단하는 것이 아니라, 카드가 자신 스스로 판단할 수 있도록 로직을 수정했습니다.

laco-dev commented on src/main/kotlin/blackjack/model/Cards.kt:
```diff
@@ -0,0 +1,33 @@
+        _value.add(card)
+    }
+
+    fun isBlackjack(firstTurn: Boolean): Boolean = CardsStatus.from(calculateScore(), firstTurn) == CardsStatus.BLACKJACK
```
블랙잭 상태는, 첫번째 턴이라기 보다는 "카드가 2장일 때 합계가 21점이면" 으로 접근을 하시는 방향이 좋을거에요!

wondroid-world replied:
페로로..... 페로로처럼 생각하면, 더 객체 지향적인 코드가 되겠어요......(조금 많이 감동받았습니다....*)

[ 관련 커밋 ] eff2d3ecb8818f4a7798c2f0d5d52906b6fdf1fc

laco-dev commented on src/main/kotlin/blackjack/model/Cards.kt:
```diff
@@ -0,0 +1,33 @@
+
+    fun isBust(): Boolean = CardsStatus.from(calculateScore()) == CardsStatus.BUST
+
+    fun calculateScore(): Int {
+        val aceCount: Int = value.count { card -> card.isDenominationAce(card) }
+        val score: Int = value.sumOf { card -> card.denomination.number }
+        return when (aceCount) {
+            1 -> if (score < 11) score + 11 else score + 1
+            2 -> if (score < 10) score + 11 + 1 else score + aceCount * 1
+            3 -> if (score < 9) score + 11 + 1 + 1 else score + aceCount * 1
+            4 -> if (score < 8) score + 11 + 1 + 1 + 1 else score + aceCount * 1
+            else -> score
+        }
+    }
```
알고리즘적이 조금 복잡하게 느껴지는데 쉽게 생각해보세요!

> 카드뭉치에 에이스 카드가 있을 때 .... 보너스 점수 10점을 획득한다.

wondroid-world replied:
페로로의 힌트 덕분에 손쉽게 수정할 수 있었습니다!

[ 관련 커밋 ] eb090855f5908671a4bd85272c14389bdd462900

laco-dev commented on src/main/kotlin/blackjack/model/GameResult.kt:
```diff
@@ -0,0 +1,30 @@
+    ;
+
+    companion object {
+        fun of(
+            standardScore: Int,
+            comparedScore: Int,
+        ): GameResult {
+            when {
+                standardScore == comparedScore -> return PUSH
+                standardScore > comparedScore -> return WIN
+            }
+            return LOSE
+        }
```
블랙잭 규칙으로 보면, 21점이 넘는다거나 하면 무조건 패배가 됩니다.

참가자 입장에서 승부가 벌어지고 있는데 단순히 점수만으로 추가로 결정하기에는 로직이 분산되어 있는 느낌이 들기도 합니다.
가령 점수가 높더라도 질 수도 있으니까요.

wondroid-world replied:
페로로! 저 블랙잭 구조를 아예 손을 보고 있어요! 이부분은 다음번 리뷰 요청때, 반영하겠습니다!

laco-dev commented on src/main/kotlin/blackjack/model/Player.kt:
```diff
@@ -0,0 +1,39 @@
+package blackjack.model
+
+open class Player(
+    open val name: String,
+    open val cards: Cards = Cards(emptyList()),
+) {
```
open class, abstract class, interface의 차이점을 알고 계신가요?
플레이어 그리고 딜러는 어떤 관계로 나타내야 적절할까요?

가령 딜러와 플레이어는 동일한 로직을 가지지만 일부는 서로 다른 로직을 수행하고는 합니다.
이를 표현하기 위한 적절한 방법은 무엇일까요?
아래 힌트를 참고해보세요 :)
```kotlin
#1
Participant
⨽Player
⨽Dealer

#2
Player
⨽Dealer
```

wondroid-world replied:
1. **Player와 Dealer는 Participant를 상속 받는다.**

[ 관련 커밋 ]
525b463a2b8ba0b72f6a52a4e947966b77f83a27

[ 이유 ]

페로로 저는 open class, abstract class, interface의 차이점을 생각해봤습니다.

open class : 객체 생성 가능, 상속 가능, 생성자 상속 가능, 내가 오픈한 메서드나 속성에 한해서 override 가능
abstract class : 객체 생성 불가능, 상속 가능, abstract 메서드에 대해서 override 무조건 오버라이드 해야함
interface : 공통 메서드를 정리해 놓은 곳

이 중에서 Participant라는 공통  추상 클래스를 생성하고 이를 각각 상속받는 것으로 결정을 내렸습니다. 왜냐하면, participant라는 객체는 생성이 가능하지 않아야합니다. 그리고 공통 property와 메서드를 상속 받아야 하기 때문에 abstract class로 만들었습니다. 그리고 dealer와 player 공통으로 가지는 메서드는 abstract class에서 생성을 했습니다.

laco-dev commented on src/main/kotlin/blackjack/model/Player.kt:
```diff
@@ -0,0 +1,39 @@
+    override fun hashCode(): Int = name.hashCode()
+
+    companion object {
+        fun from(name: String): Player {
+            require(name != "딜러") { "플레이어는 딜러라는 이름을 가질 수 없습니다." }
+            require(name.length in 1..5) { "플레이어는 1에서 5사이 길이의 이름만 가질 수 있습니다." }
+            return Player(name)
+        }
```
한편 지금 상속 구조라면 딜러 == 플레이어인데 플레이어가 딜러 이름을 가질 수 없다는 점이 매우 모순적으로 느껴지기는 합니다...

wondroid-world replied:
이 부분은 abstract class를 상속받아서, player에서만 검증하도록 할 수 있을 거 같습니다!

[ 관련 커밋 ] 4bf253e1a5d2a50581ec44b583192dcde3a4c471

laco-dev reviewed (changes requested):
1단계 고생 많으셨어요.

여러가지 고민과 미래를 그리는 모습이 보이는데요!
기본적인 개념과 연습이 충분히 된 다음에 고민을 해야 할 요소들로 보여요.
아직은 잘 모르는데도 불구하고 오버 엔지니어링을 하려는 것으로 느껴지기도 합니다.

점진적으로 개선해볼 요소들이 있으니 우선은 관심사 분리부터 연습해 보도록 하시죠.

wondroid-world committed:
refactor: 카드 모양의 koreanName 삭제

- 카드 모양의 koreanName은 view에서 판단할 로직이라고 생각함. 관심사 분리를 위해서, koreanName property 삭제

wondroid-world committed:
refactor: Denomination의 title 삭제

- domain이 view의 의존 관계를 가지고 있어 관심사를 분리함

wondroid-world committed:
refactor: cards의 getCardsInformaiton 삭제

- 도메인이 view에 의존적이여서 삭제

wondroid-world committed:
reafactor: isDenominationAce 파라미터 삭제와 함수명 변경

- 자신의 끗수가 Ace인지 판단하는 로직임. 이를 판단하기 위해서, 파라미터가 필요없으므로 삭제
- Ace인지 알기 쉬운 메서드명으로 변경함
- isAce 메서드의 test 추가

wondroid-world committed:
refactor: PlayerBehavior의 answer 필드 삭제와 HIT, STAY 구분 로직 inputview로 변경

wondroid-world committed:
refactor: 딜러 스스로가 hit or stay할지 판별

wondroid-world committed:
refactor: 카드 스스로 상태를 판단하도록 로직 수정

- 카드 상태에서 카드들의 상태를 판단하는 로직 삭제
- 카드들 스스로 자신의 상태를 판단하도록 로직 생성
- player에서도 card들의 상태를 판단하는 로직 삭제

wondroid-world committed:
refactor: 블랙잭 판단 로직 수정

- 첫 턴인지가 아니라, 카드의 사이즈가 2이고 합이 21일때, 블랙잭을 반환하도록 수정

wondroid-world committed:
refactor: Player와 Dealer는 Participant를 상속

wondroid-world committed:
refactor: 패키지 구조 변경

wondroid-world committed:
refactor: 게임 결과 UI와 도메인 로직 분리

wondroid-world committed:
refactor: addCard 접근 제어자 변경

- protected -> public으로 변경

wondroid-world committed:
feat: 이름 검증 로직 추가

wondroid-world committed:
refactor: Cards의 필드인 size 삭제

- test를 위한 필드임으로 삭제

wondroid-world committed:
test: 게임 결과에 대한 도메인과 ui 분리

wondroid-world committed:
refactor: 점수 계산 로직 변경

wondroid-world commented:
페로로! 안녕하세용!

> 블랙잭 규칙으로 보면, 21점이 넘는다거나 하면 무조건 패배가 됩니다. 참가자 입장에서 승부가 벌어지고 있는데 단순히 점수만으로 추가로 결정하기에는 로직이 분산되어 있는 느낌이 들기도 합니다.가령 점수가 높더라도 질 수도 있으니까요.

이 부분은 다음 리뷰 때, 반영하고 싶었습니다. 왜냐하면 지금 대대적으로 구조 변경을 하고 있어요! controller에서 너무 많은 일을 하고 있어서 객체로 분리하고 싶었습니다. 그리고 중복된 로직도 너무 많다고 생각했습니다! 

현재, 페로로가 리뷰 남겨준 부분과 페어 리뷰어가 남겨준 리뷰를 바탕으로 수정을 했습니다. 페로로가 일단 관심사를 분리해야 된다라고 해주셔서 도메인에 view와 관련된 로직을 수정했습니다.

페로로 저는 부족한것이 많기 때문에 많이 때려주세요 부탁드려요.

laco-dev commented on src/test/kotlin/dsl/PersonBuilder.kt:
```diff
@@ -0,0 +1,32 @@
+    private val hobbies = mutableListOf<String>()
+    private val skills = mutableListOf<Skill>()
+
+    fun name(value: String) {
+        name = value
+    }
+
+    fun mbti(value: String) {
+        mbti = value
+    }
```
Builder의 특징은 선택적으로 속성을 부여하는 것 일텐데, name을 호출하지 않고 build() 하면 어떤 문제가 생기나요?

어떻게 해결해야 할까요?
"호출하지 않으면 이름은 빈값이다" 라는 답변은 아니길 바랍니다..

힌트는 이름은 필수라는 점입니다.

laco-dev commented on src/test/kotlin/dsl/CatBuilder.kt:
```diff
@@ -0,0 +1,19 @@
+        gender = value
+    }
+
+    fun build(name: String): Cat {
+        return Cat(name, age, gender)
+    }
```
요기는 대응을 해둔 것 같은데... build 함수에 파라미터로 넘기는게 맞나요? 아니면 또 다른 방법이 있나요?

laco-dev commented on src/main/kotlin/blackjack/model/card/Cards.kt:
```diff
@@ -0,0 +1,27 @@
+class Cards(
+    value: List<Card>,
+) {
+    private val _value: MutableList<Card> = value.toMutableList()
+    val value: List<Card> get() = _value.map { card -> card.copy() }
```
> 일차적으로 mutable 주소값을 노출하지 않기위해서, 딥카피를 해서, 값을 전달하고 싶었습니다!

Card가 이미 불변 객체인데 노출이 되어서 문제가 될 점이 있을까요?

laco-dev commented on src/main/kotlin/blackjack/model/card/Cards.kt:
```diff
@@ -0,0 +1,27 @@
+        return score
+    }
+
+    fun getStatus(): CardsStatus {
+        val cardsScore = calculateScore()
+        if (_value.size == 2 && cardsScore == 21) return CardsStatus.BLACKJACK
+        if (cardsScore > 21) return CardsStatus.BUST
+        return CardsStatus.NONE
+    }
```
좋습니다 :)
반복되는 21점에 대한 값만 상수화 해주면 좋겠어요

laco-dev commented on src/test/kotlin/blackjack/model/CardsTest.kt:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Test
+    fun `처음 턴이고, 카드들의 합이 21이면 CardsStatus의 BLACKJACK 값을 반환한다`() {
```
첫 턴이라는 관심사가 없어졌기 때문에 테스트 케이스도 맞춰서 수정되어야 할 것 같네요

laco-dev commented on src/test/kotlin/blackjack/model/DenominationTest.kt:
```diff
@@ -0,0 +1,31 @@
+import org.junit.jupiter.params.provider.ValueSource
+
+class DenominationTest {
+    @ParameterizedTest
+    @ValueSource(strings = ["JACK", "QUEEN", "KING"])
+    fun `끗수가 J,Q,K일 경우 숫자가 10이다`(denomination: Denomination) {
+        assertThat(denomination.number).isEqualTo(10)
+    }
```
ParameterizedTest를 활용한다면 테스트 이름을 이렇게도 작성해 볼 수 있을거에요
> 끗수가 {0}일 경우 숫자가 10이다.

laco-dev commented on src/main/kotlin/blackjack/model/participant/Player.kt:
```diff
@@ -0,0 +1,20 @@
+        require(name.length in 1..5) { "플레이어는 1에서 5사이 길이의 이름만 가질 수 있습니다." }
+    }
+
+    lateinit var result: GameResult
+
+    fun updateResult(dealerGameResult: GameResult) {
+        result = GameResult.reversed(dealerGameResult)
+    }
```
이쪽은 구조를 바꾸신다고 하여 의견만 남겨둡니다.

플레이어가 딜러 결과를 통해 내 결과를 업데이트 한다는 점이 꽤나 어색하게 느껴지고, 재사용이 어렵습니다.
승부를 벌이는 것은 참가자와 참가자간의 승부로 일반화를 할 수 있으려나요?

1. 딜러 vs 플레이어
2. 플레이어 vs 플레이어
3. 참가자 vs 참가자

laco-dev commented on src/main/kotlin/blackjack/model/participant/Players.kt:
```diff
@@ -0,0 +1,13 @@
+        require(value.size == value.toSet().size) { "플레이어 이름은 중복될 수 없습니다." }
+    }
+
+    fun findBlackjackPlayer(): List<Player> = value.filter { player -> player.cards.getStatus() == CardsStatus.BLACKJACK }
```
결과를 출력하는 과정에서. 플레이어의 상태가 블랙잭인지를 판단하도록 할 수 있지 않나요?
블랙잭 플레이어들을 찾는 과정이 필요할지 궁금해서요.

이 또한 플레이어에게 메시지를 전달하도록 만들어도 되겠구요

laco-dev commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,87 @@
+        players.value.forEach { player ->
+            println(
+                "${player.name}카드: ${
+                    player.cards.value.map { it.shape.toString() + it.denomination }.joinToString(", ")
```
이 부분 너무 반복되는데 개선 해보면 좋을 것 같아요.

laco-dev commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,87 @@
+        println("게임에 참여할 사람의 이름을 입력하세요.(쉼표 기준으로 분리)")
+    }
+
+    fun printPlayersCards(
+        dealer: Dealer,
+        players: List<Player>,
+    ) {
+        val playersNames: String = players.joinToString(", ") { it.name }
+        println("\n${dealer.name}와 ${playersNames}에게 2장의 카드를 나누었습니다.")
+        println(
+            "${dealer.name}: ${
+                dealer.cards.value
+                    .map { it.shape.toString() + it.denomination }
+                    .joinToString(", ")
+            }",
+        )
+        players.forEach { player ->
+            printPlayerCard(player)
+        }
+        println()
+    }
```
최초에 딜러는 1장의 카드를, 플레이어는 2장의 카드를 공개한다 라는 규칙이 있을거에요.
이 규칙은 명확한 비즈니스 로직인데 어떻게 구현할지 추상화 관점에서 접근해보면 도움이 될거에요

laco-dev commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,170 @@
+        playGames(players, dealer)
+    }
+
+    private fun getCards(
+        players: Players,
+        dealer: Dealer,
+    ) {
+        repeat(2) {
+            getCardsToPlayer(players.value)
+            getCardsToDealer(dealer)
+        }
+        outputView.printPlayersCards(dealer, players.value)
+    }
```
생성자로 준비한 Cards를 테스트에서만 쓰이다보니, 처음 플레이어와 딜러를 준비 할 때 2장의 카드를 같이 만들어서 활용할 수 있지 않을까요?

또한 이 함수이름이 getXXX 형태인데 반해 반환도 하지 않고 상태를 바꾸기 때문에 함수 이름이 적절하지 않습니다.

laco-dev commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,170 @@
+        outputView.printPlayersCards(dealer, players.value)
+    }
+
+    private fun getCardsToPlayer(players: List<Player>) {
+        players.forEach { player ->
+            val card = cardDeck.pickCard()
+            player.addCard(card)
+        }
+    }
+
+    private fun getCardsToDealer(dealer: Dealer) {
+        val card = cardDeck.pickCard()
+        dealer.addCard(card)
+    }
```
비슷하게 함수 이름이 어색합니다!

laco-dev commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,170 @@
+    }
+
+    private fun executePlayerGameLogic(player: Player) {
+        while (player.cards.getStatus() != CardsStatus.BUST) {
+            outputView.printPlayerBehaviorGuide(player)
+            val playerBehavior: PlayerBehavior = inputView.readPlayerBehavior()
+
+            if (executePlayerBehavior(playerBehavior, player)) break
+        }
```
사실 Hit는 상태라기 보다는 더 뽑껬다고 행동한 액션에 가까워요.
그런 의미에서 본다면 딜러가 카드를 더 뽑을 수 있는지 여부를 뜻하는 isHit() 대신 canDraw()와 같이 접근하면 플레이어도 동일한 역할을 부여할 수 있습니다.

laco-dev commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+    private fun executePlayerBehavior(
+        playerBehavior: PlayerBehavior,
+        player: Player,
+    ): Boolean {
+        when (playerBehavior) {
+            PlayerBehavior.HIT -> {
+                player.addCard(cardDeck.pickCard())
+                outputView.printPlayerCard(player)
+                if (isPlayerBust(player)) return true
+            }
+
+            PlayerBehavior.STAY -> return true
+        }
+        return false
+    }
```
execute가 Boolean을 반환하면 어떤 값인지 사용하는 입장에서 예측이 될까요?

wondroid-world replied:
맞아요 페로로 이부분은 함수명에서도 유츄가 안되어서 판단하기 힘들거 같아요!

laco-dev commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,170 @@
+        return false
+    }
+
+    private fun isPlayerBust(player: Player): Boolean {
+        if (player.cards.getStatus() == CardsStatus.BUST) {
+            outputView.printBust(player)
+            return true
+        }
+        return false
+    }
```
플레이어가 버스트인지 확인하는데 출력까지 2가지 일을 하고 있지 않을까요?

laco-dev commented on src/main/kotlin/blackjack/controller/BlackjackController.kt:
```diff
@@ -0,0 +1,170 @@
+) {
+    private val cardDeck = CardDeck(CardsMaker.CARDS)
+
+    fun run() {
+        outputView.printStartMessage()
+        val players: Players = inputView.readPlayers()
+        val dealer = Dealer()
+
+        getCards(players, dealer)
+        playGames(players, dealer)
+    }
```
전반적으로 컨트롤러에서 게임에 대한 비즈니스로직을 처리하고 있어요.
만약 지금 구조에서는 컨트롤러 없이 게임을 진행할 수 있을까요?

꽤나 어려울 수 있는 주제이지만 게임을 진행하기 위한 도메인 모델을 고려해볼 수 있겠습니다.

laco-dev commented on src/main/kotlin/blackjack/model/participant/Dealer.kt:
```diff
@@ -0,0 +1,24 @@
+        return dealerScore <= 16
+    }
+
+    fun updateResult(playerScore: Int): GameResult {
+        val dealerScore: Int = cards.calculateScore()
+        val result: GameResult = GameResult.of(dealerScore, playerScore)
+        _results[result] = _results.getOrDefault(result, 0) + 1
+        return result
+    }
```
아까 적었던 의견인데 결국 승부를 한다는 행위가 게임 결과를 반환하기만 하면 되는데 굳이 상태까지 업데이트 하고 있을 이유가 있나? 싶네요.

laco-dev reviewed (approved):
대대적으로 구조를 수정하고 계시다 하여 1단계에서 너무 지체될까봐 다음 단계에서 같이 확인하려고 합니다.

남겨드린 내용 중 필요한 수정사항은 같이 챙겨주시고, 아닌 점은 참고만 해주시면 되겠습니당.

저는 크루를 때리지 않아요... 😭
